### DISCUSSION TITLE
Interaction on the backend

### DISCUSSION THREAD

**Discussion by JRII972** [UniTime/unitime#158] [Q&A] [answered]
Hi,

I'm creating a management tool for my company, and we need a solver.
I would need to link my tools (web app) and Unitime, to avoid having to rewrite certain information.

Apart from import export via XML, is there a way to interact directly by programming since they are on the same server.
I looked through the documentation and didn't see any. And I don't really want to hit directly the database.

I think, 2 month ago when a search on the web, i might see a way by calling a url, but i don't find it. Does it exist ?

Thank you by advance

**Comment by tomas-muller** [comment]
If you just need a solver, you do not need to use UniTime, just the CPSolver library. See [https://github.com/UniTime/cpsolver](https://github.com/UniTime/cpsolver). It contains the solver framework (see org.cpsolver.ifs package), a few examples (org.cpsolver.ifs.examples, see [Constraint Solver Examples](https://www.cpsolver.org/cpsolver_examples.php)), and a course timetabling, examination timetabling, student scheduling and instructor scheduling solvers that UniTime is using. If you are interested in the course timetabling solver, for example, there is also a command line interface (see the [org.cpsolver.coursett.Test](https://github.com/UniTime/cpsolver/blob/master/src/org/cpsolver/coursett/Test.java) class).

If you want to interface with UniTime, we do have a set of APIs available, see [bit.ly/unitime44api](http://bit.ly/unitime44api) for more details.

**Reply by JRII972** [reply]
Thx, i'll implement the CPSolver library later, and use UniTime APIs for saving time.

Thank you

**Reply by JRII972** [reply]
I see the only way to add common data using a script?

I've read the API code, and some don't have a post implemented, others require a script in the settings.

Do you have an example script that injects data?

**Reply by tomas-muller** [reply, marked as answer]
Yes, most of the APIs were created to get data from UniTime, though there are some that can make changes (e.g., Rooms API -- see the [RoomsConnector](https://github.com/UniTime/unitime/blob/master/JavaSource/org/unitime/timetable/api/connectors/RoomsConnector.java) for the implementation). They are usually added to fulfill a certain business need (they are not meant to cover the whole UniTime data model). You can also use one of our [XML interfaces](https://www.unitime.org/uct_interfaces.php) and use the Data Exchange API to import the file.



See the [Instructor Assignment Imports](https://github.com/UniTime/unitime/blob/master/Documentation/Scripts/Instructor%20Assignment%20Imports.xml) scripts as an example of a script that imports data. There are two scripts, both taking a CSV file and importing instructors and their teaching preferences and teaching requests respectively for the instructor scheduling module (see the [Instructor Scheduling Manual](https://docs.google.com/document/d/1JksFtB9ecIoKpuyY0SDdl3dKnATe8UQAPKlGfBWi2JU/edit#heading=h.kakomxlphxc0) for more details).
